[MRG+1] Isotonic calibration

agramfort opened the pull request:
calibration module with platt and Isotonic calibration.

A few issues :
- [x]  nice calibration plots
- [x]  add tests for metrics
- [x]  add narrative doc
- [x] speed up the test execution as much as possible
- [x] fix multi-class support for the isotonic regression method

jmetzen commented:
I've sped up the test execution as much as possible; going to even smaller datasets or lower values for the cv parameter would result in a setting where calibration does no longer provide reliable improvements. But tests are now considerably faster already.

Multi-class support for the isotonic method was already working since PR #3995  was merged; I've updated the tests such that they alway test both methods. I also extended the documentation such that cover the multiclass case.

I cannot update the todo-list and the title of this PR, but from my point of view, the todos have been addressed and we could change the title to [MRG]

ogrisel commented:
@jmetzen I just fetched the current state of your branch and the calibration curve example seems broken now:

![calibration_plot](https://cloud.githubusercontent.com/assets/89061/5613845/8e28cdf2-94ea-11e4-9c4c-11ec864bf6de.png)

```
Logistic:
	Brier: 0.166
	Precision: 0.725
	Recall: 0.696
	F1: 0.711
Naive Bayes:
	Brier: 0.480
	Precision: 0.465
	Recall: 0.978
	F1: 0.631
Naive Bayes + Isotonic:
	Brier: 0.281
/volatile/ogrisel/code/scikit-learn/sklearn/metrics/classification.py:936: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)
	Precision: 0.000
	Recall: 0.000
/volatile/ogrisel/code/scikit-learn/sklearn/metrics/classification.py:936: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)
	F1: 0.000
Naive Bayes + Sigmoid:
	Brier: 0.230
	Precision: 0.000
	Recall: 0.000
	F1: 0.000
```

Do you observer the same behavior (no points for the lower bins of the calibrated models)?

All the tests pass though.

ogrisel commented:
BTW thanks for the test speedup, on my box the 5 calibration tests run in 400ms.

jmetzen commented:
@ogrisel, you are right, there is something strange going on here. The results for "isotonic" have changed compared to the graphic inlined by @agramfort [above](https://cloud.githubusercontent.com/assets/161052/3791869/82ae805a-1b57-11e4-93ea-d261e566170d.png), otherwise it is unchanged (note that everything is transposed since the axes had been interchanged originally).

The reason that isotonic has changed are the changes merged in PR #3995, which prevent that nan occurs in isotonic_regression.py. Strangely, the calibration curve was better before the bug was fixed. One possible reason might be that the bug originally resulted in ignoring all instances during calibration for which the base classifier returned probability 0, which might have improved things. I will investigate this issue in more detail later and see if I can fix the example.

ogrisel commented:
@jmetzen I am investigating with changing this example to use a subset of covertype with gradient boosting classifier or a linear svm. I will let you know about my findings.

ogrisel commented:
I have not found the time to refactoring the calibration plot example and won't have time to do it before next week.  Feel free to work on that in the mean time if you wish.

amueller commented on sklearn/calibration.py:
```diff
@@ -0,0 +1,489 @@
+        The method to use for calibration. Can be 'sigmoid' which
+        corresponds to Platt's method or 'isotonic' which is a
+        non-parameteric approach. It is not advised to use isotonic calibration
+        with too few calibration samples (<<1000) since it tends overfit. Use 
```
tends to overfit

amueller commented:
I'm unsure about the multiclass plot.
What is the message it conveys? That all predictions got more uncertain and that blue got more uncertain than the rest? Also, I am a bit doubtful about the result. This is sigmoid calibration, right? So
it is a function from the simplex to the simplex using 3 parameters and a sigmoid. That should be a really smooth function, right? But there seem to be some very erratic arrows. Shouldn't that be a smooth vector field?

amueller commented on sklearn/calibration.py:
```diff
@@ -0,0 +1,489 @@
+        Specific cross-validation objects can be passed, see
+        sklearn.cross_validation module for the list of possible objects.
+        If "prefit" is passed, it is assumed that base_estimator has been
+        fitted already and all data is used for calibration.
```
Attributes are not documented!

jmetzen replied:
done

amueller commented on sklearn/calibration.py:
```diff
@@ -0,0 +1,489 @@
+            cv = _check_cv(self.cv, X, y, classifier=True)
+            for train, test in cv:
+                this_estimator = clone(self.base_estimator)
+                if sample_weight is not None and \
```
Why is it silently ignored if it is not in argspec? because it is used in the calibration? seems slightly magic, but not sure what to do about it.

jmetzen replied:
this is supposed to deal with the case that there are sample-weights which can be used in calibration but the base-estimator does not support sample-weights. We could warn in this situation,

ogrisel replied:
Indeed I will add a warning and a test before merging.

amueller commented:
There are tabs in the code (and a lot of trailing spaces). Please run pep8.

amueller commented:
I guess I understand now a bit better what is happening in the multi-class example. The calibrated random forest is an ensemble of random forests that are completely independent of the first fit.
In this case, I feel the arrows are misleading, as they don't really show what the calibration does. These are all different random forests with different seeds and anything can move anywhere basically. Only in aggregate can you say that calibration happens.

I think it would be nicer to have an image like this:
![multi_class_calibration](https://cloud.githubusercontent.com/assets/449558/5745262/7695faae-9bf5-11e4-9d76-797937233908.png)
Here I used ``cv='prefit'`` and a validation set. It is a bit unfair to compare then the loss of the calibrated and uncalibrated classifiers, as the uncalibrated didn't make use of the validation set. But you can see the effect of calibration for a given trained classifier much more clearly, I think.

amueller commented:
Btw, here is how the whole calibration function looks like on the single estimator. I guess it is not as smooth as I imagined:
```python
calibration = sig_clf.calibrated_classifiers_[0]
calibration.calibrators_[0]
p1d = np.linspace(0, 1, 20)
p0, p1 = np.meshgrid(p1d, p1d)
p2 = 1 - p0 - p1
p = np.c_[p0.ravel(), p1.ravel(), p2.ravel()]
p = p[p[:, 2] >= 0]
prediction = np.vstack([calibrator.predict(this_p) for calibrator, this_p in zip(calibration.calibrators_, p.T)]).T

for i in range(prediction.shape[0]):
    plt.arrow(p[i, 0], p[i, 1],
              prediction[i, 0] - p[i, 0],
              prediction[i, 1] - p[i, 1], head_width=1e-2, color=colors[np.argmax(p[i])])
```
![multi_class_calibration_vectorfield](https://cloud.githubusercontent.com/assets/449558/5745933/22ebbff6-9bfa-11e4-9ae8-96eb4564ae23.png)

jmetzen commented:
@amueller I agree with what you said regarding the multi-class example. I will revise the example accordingly

jmetzen commented:
@amueller I modified the multiclass example according to your comments. The illustration of the calibration function looks a bit differently since one also has to renormalize such that the three probabilities sum to 1:

![plot_calibration_multiclass_0011](https://cloud.githubusercontent.com/assets/1116263/5792413/53951100-9f17-11e4-99b5-397d0788ae8b.png)

@ogrisel The calibration curve example still needs revision; I will work on it during the next days.

amueller commented:
Huh, that does look more like what I was expecting, thanks.
Not sure if we want to visualize the whole area or just the data points (both in one plot is probably too much, right?) I do think it is a nice visualization of the rescaling, though :)

jmetzen commented:
I think it makes sense to have both visualizations; I have also adapted calibration.rst accordingly.

jmetzen commented:
@ogrisel I've revised the calibration curve example:
![plot_calibration_curve_0011](https://cloud.githubusercontent.com/assets/1116263/5822652/3356204e-a0d6-11e4-849a-4b50460a9e7e.png)

Instead of the covertype it is now based on an artificial binary classification dataset with redundant features. I believe it illustrates the properties of the calibration methods nicely (documentation is updated); but I am also open for suggestions.

I also encountered a further issue of isotonic_regression which occurs when multiple inputs are identical but have different outputs. As a workaround, I added a small random jitter to all inputs.

ogrisel commented:
This is very good. Could you please add a second similar plot with LogisticRegression / SVC(kernel='linear') / SVC(kernel='linear') + Platt scaling  / SVC(kernel='linear') + IC?

On a second plot to avoid overloading the existing plot. Then explain in the text that NB and SVC are both broken but in opposite ways: the margin of the hinge loss of SVC will make the model focus on hard samples, close to the decision boundary (support vectors) and therefore lead to an under-confident model while NB is generally overconfident in the presence of correlated features.

jmetzen commented:
I've added the LinearSVC example and updated the text accordingly:

![plot_calibration_curve_0021](https://cloud.githubusercontent.com/assets/1116263/5988240/39a59838-a957-11e4-97ae-e99645e6e832.png)

ogrisel commented on doc/modules/calibration.rst:
```diff
@@ -0,0 +1,190 @@
+of the predict_proba method can be directly interpreted as a confidence level.
+For instance, a well calibrated (binary) classifier should classify the samples
+such that among the samples to which it gave a predict_proba value close to 0.8,
+approx. 80% actually belong to the positive class. The following plot compares
```
approximately

ogrisel commented on doc/modules/calibration.rst:
```diff
@@ -0,0 +1,190 @@
+   :align: center
+
+.. currentmodule:: sklearn.linear_model
+:class:`LogisticRegression` returns well calibrated predictions as it directly
```
... calibrated predictions by default ...

ogrisel commented on examples/plot_compare_calibration.py:
```diff
@@ -0,0 +1,119 @@
+
+* Support Vector Classification (SVC) shows a similar sigmoid curve as
+the  RandomForestClassifier, which is typical for maximum-margin methods
+(compare Niculescu-Mizil and Caruana [1]).
```
I think this is confusing: `SVC(probability=True)` is already using the internal Platt-scaling implemented in libsvm which is supposed to attenuate the margin effect. I think we should either add a note here to state that `SVC(probability=True)` are already calibrated but this does not seem to work correctly on this data, or use the uncalibrated decision function of `LinearSVC` to illustrate this point in this example.

ogrisel commented on sklearn/tests/test_calibration.py:
```diff
@@ -0,0 +1,180 @@
+        probas = cal_clf.predict_proba(X_test)
+        assert_array_almost_equal(np.sum(probas, axis=1), np.ones(len(X_test)))
+        assert_true(clf.fit(X_train, y_train).score(X_test, y_test) <=
+                    cal_clf.fit(X_train, y_train).score(X_test, y_test))
```
It's better to use `assert_greater` to get informative error messages without having to launch a debugger when stuff breaks on travis.

ogrisel replied:
Actually, I am not sure that this assertion makes sense in the first place. Why would probability scaling improve the hard prediction score in the first place? The `score` method of classifiers returns the accuracy (that is the opposite of the zero-one loss). AFAIK it should not necessarily be improved by probability calibration: the location of the decision threshold might move a bit but I see no principled reason why probability would necessarily improve the exact location of the decision boundary w.r.t. the accuracy score.

ogrisel commented on sklearn/tests/test_calibration.py:
```diff
@@ -0,0 +1,180 @@
+        cal_clf.fit(X_train, y_train)
+        cal_clf_probs = cal_clf.predict_proba(X_test)
+        cal_score = log_loss(y_test, cal_clf_probs)
+        assert_true(cal_score < score)
```
It would be interesting to test the `log_loss` of a calibrated `LinearSVC` vs the `log_loss` of a naive scaling of the `decision_function` to the `[0, 1]` range.

ogrisel commented on sklearn/tests/test_calibration.py:
```diff
@@ -0,0 +1,180 @@
+def test_sigmoid_calibration():
+    """Test calibration values with Platt sigmoid model"""
+    exF = np.array([5, -4, 1.0])
+    exY = np.array([1, -1, -1])
```
I don't understand what those variables represent. What does `ex` mean?

jmetzen replied:
No idea, this test was probably written by @agramfort before I joined working on this PR

ogrisel commented on sklearn/tests/test_calibration.py:
```diff
@@ -0,0 +1,180 @@
+    exF = np.array([5, -4, 1.0])
+    exY = np.array([1, -1, -1])
+    # computed from my python port of the C++ code in LibSVM
+    AB_lin_libsvm = np.array([-0.20261354391187855, 0.65236314980010512])
```
What does `AB_lin` mean?

ogrisel commented on examples/plot_calibration.py:
```diff
@@ -0,0 +1,107 @@
+print("No calibration: %1.3f" % clf_score)
+
+clf_isotonic_score = brier_score_loss(y_test, prob_pos_isotonic, sw_test)
+print("With isotonic calibration: %1.3f" % clf_isotonic_score)
```
I get the following error when running this example:

```
Traceback (most recent call last):
  File "examples/plot_calibration.py", line 71, in <module>
    clf_isotonic_score = brier_score_loss(y_test, prob_pos_isotonic, sw_test)
  File "/Users/ogrisel/code/scikit-learn/sklearn/metrics/classification.py", line 1606, in brier_score_loss
    y_true = _check_binary_probabilistic_predictions(y_true, y_prob)
  File "/Users/ogrisel/code/scikit-learn/sklearn/metrics/classification.py", line 1529, in _check_binary_probabilistic_predictions
    raise ValueError("y_prob contains values greater than 1.")
ValueError: y_prob contains values greater than 1.
```

ogrisel replied:
With the debugger I found that:

```
(pdb) p y_prob.max()
1.0000000000000002
```

ogrisel replied:
It seems that this regression was introduced by: #4111 (the use of `kind="slinear"` for `interp1d` in the isotonic regression model).

I think we should make the Brier score ignore such small deviance.

amueller replied:
Yeah that introduced some issues, and there already where some issues before :-/ see #4185.

amueller replied:
Also #2507.

ogrisel commented:
I think we should move all the calibration example into a a specific folder such as `example/calibration/` to avoid crowding the root example folder too much. The links in the documentation will have to be updated.

ogrisel commented:
The new multiclass example is very good.

ogrisel commented on examples/plot_calibration_curve.py:
The Sigmoid calibration does not seem to work as expected on the inverted sigmoid shape of NB models. Is this an intrinsic limitation of the parametrization of the Platt method that expects a non-inverted sigmoid or a limitation of our own implementation? Based on my understanding of Platt's method I would think the former (that is a limitation of the parametrization of the method that is not suitable for NB models). If this is correct I think we should rewrite that paragraph to make this limitation more explicit.

ogrisel commented on examples/plot_calibration_curve.py:
Changing the ordering of this paragraph with the one on the calibration of the SVM model might help makes things clearer: AFAIK Platt's sigmoid scaling was devised for margin-based models in the first place for which the normalized decision function exhibit the expected (non-inverted) sigmoid shape.

ogrisel commented:
PS: this PR need another rebase on top of master. Sorry for the slow review latency.

ogrisel commented on doc/modules/calibration.rst:
I would rather say that this can be attributed to the fact that the parametric assumption made by the sigmoid method does not fit the empirical shape of the typical reliabity curve of naive Bayes models.

jmetzen commented:
Thanks for your review @ogrisel. I've addressed most of your comments. What's still open:
 * You suggested that brier_score_loss should not complain about y_prob values that minimally exceed 1.0. This behavior actually stems from _check_binary_probabilistic_predictions in metrics/classification.py. I am not sure if we should change it there. I've modified CalibratedClassifierCV instead such that it checks the output and replaces any 1.0+eps by 1.0
 * "It would be interesting to test the log_loss of a calibrated LinearSVC vs the log_loss of a naive scaling of the decision_function to the [0, 1] range.": I am not sure how to get from decision_function's output to a value in [0,1] in the non-binary setting. Is there a canonical way?
 * Regarding using the accuracy in test_calibration_multiclass: I had the same thoughts as you that the accuracy is only very indirectly connected to the underlying probability values. log-loss would be more appropriate but it requires to get from decision function to the [0, 1] range (see above)

amueller commented:
For the second question: The canonical way to get probabilities from decision values is softmax, right? However that is not what the OVR logistic regression does (which does a sigmoid per class and then renormalizes iirc).

jmetzen commented:
Softmax based on the decision scores of different OVR SVCs would be a bad idea in general in my opinion as the decision scores are incommensurable. But I guess for the sake of this test it should be ok since the right way to do it would be Platt's calibration, which is escatly the point of this test.

jmetzen commented:
As for the travis failure: the issue here is that we have a dataset with 4 labels and 2 of these occur only twice. Thus, in the 3-fold startiefied crossvalidation used internally in CalibratedClassifierCV, we have have one fold where two of the classes don't occur in the data used for calibration. There are at least two different ways of dealing with that:
 * Raising an exception when there is too few data per class, i.e., n_examples_per_class < n_cv_folds
 * Calibrating on the data which was used for fitting the base-estimator (and warning about that)
I would prefer the second way. What do you think?

amueller commented:
I don't have a stong opinion but I am leaning towards the first way. Trying to hack a dataset with 2 samples to work seems not a good idea, and we can make the datasets just slightly larger in the tests.

ogrisel commented:
> "It would be interesting to test the log_loss of a calibrated LinearSVC vs the log_loss of a naive scaling of the decision_function to the [0, 1] range.": I am not sure how to get from decision_function's output to a value in [0,1] in the non-binary setting. Is there a canonical way?

This is only for a reliability / calibration plot so it can only be done for binary classification right? For binary classification a naive "Min Max scaling" operation will do.

ogrisel commented:
> I don't have a stong opinion but I am leaning towards the first way. Trying to hack a dataset with 2 samples to work seems not a good idea, and we can make the datasets just slightly larger in the common tests.

+1

ogrisel commented:
Actually I think that @amueller's suggestion to apply a softmax transform to the OvR decisions to turn them into (uncalibrated) probabilities that sum to 1 for each sample is reasonable:

```python
def softmax(y_pred):
    e = np.exp(-y_pred)
    return e / e.sum(axis=1).reshape(-1, 1)

y_pred = linear_svc.decision_function(X_test)
uncalibrated_log_loss = log_loss(y_test, softmax(y_pred))
```

ogrisel commented:
> You suggested that brier_score_loss should not complain about y_prob values that minimally exceed 1.0. This behavior actually stems from _check_binary_probabilistic_predictions in metrics/classification.py. I am not sure if we should change it there. I've modified CalibratedClassifierCV instead such that it checks the output and replaces any 1.0+eps by 1.0

Fine with me.

ogrisel commented:
@jmetzen could you also please rebase & squash this PR on top of master? In particular you can squash all the FIX commits that are non-informative per-se for a new feature PR.

jmetzen commented:
The remaining issues are now addressed:
  * The number of examples in the failing tests is increased (if there are too few examples, CalibratedClassifierCV raises an exception)
  * log-loss is used instead of accuracy for comparison in calibration tests (based on softmax transforming the decision-score)
  * rebased and squashed

ogrisel commented on sklearn/tests/test_calibration.py:
```diff
@@ -0,0 +1,191 @@
+        uncalibrated_log_loss = \
+            log_loss(y_test, softmax(clf.decision_function(X_test)))
+        calibrated_log_loss = log_loss(y_test, probas)
+        assert_greater_equal(uncalibrated_log_loss, calibrated_log_loss)
```
Great test :)

ogrisel commented on doc/modules/calibration.rst:
```diff
@@ -0,0 +1,195 @@
+sigmoid calibration is preferable if the calibration curve is sigmoid and when
+there is few calibration data while isotonic calibration is preferable for non-
+sigmoid calibration curves and in situations where many additional data can be
+used for calibration.
```
Very good explanation. Thanks.

ogrisel commented on examples/calibration/plot_calibration.py:
```diff
@@ -0,0 +1,107 @@
+demonstrates how to transform the decision function of a generic
+classifier into a calibrated probability. The accuracy is estimated
+with Brier's score (see http://en.wikipedia.org/wiki/Brier_score).
+
```
Please add a note here stating that in that case, only the non-parametric isotonic calibration method is able to provide a probability calibration that return probabilities close to the expected 0.5 for most of the samples belonging to the middle cluster with heterogeneous labels.

ogrisel commented on examples/calibration/plot_calibration_multiclass.py:
```diff
@@ -0,0 +1,169 @@
+prediction = \
+    np.vstack([calibrator.predict(this_p)
+               for calibrator, this_p in
+                   zip(calibrated_classifier.calibrators_, p.T)]).T
```
Please avoid using a `\` when not necessary:

```python

prediction = np.vstack([calibrator.predict(this_p)
    for calibrator, this_p
    in zip(calibrated_classifier.calibrators_, p.T)]).T
```

ogrisel commented:
@jmetzen I have submitted a PR against this branch agramfort/scikit-learn#11 to better test the calibration with categorical class labels instead of contiguous integers. (I checked the values of the log-loss and they stay approximately the same).

ogrisel commented:
Apart from my previous comments +1 on my side.

jmetzen commented:
Thanks for your comments; I changed the documentation and code accordingly and squashed

ogrisel commented:
The last commit message is no longer descriptive of its effective content: can you please either use a descriptive commit message or squash further into a bunch of very generic commits (e.g. 4 commits: one for impl, one for tests, one for examples and one for the narrative doc). Or even a single one.

jmetzen commented:
The commit message is more informative now (the commit messages of the squashed commits are also retained in the body of the commit message). I am not squashing further since this would hide that the original implementation is due to @agramfort

ogrisel commented:
> (the commit messages of the squashed commits are also retained in the body of the commit message)

I don't like that personally. I prefer to have a one-liner aggregate summary, then  a blank line and then more details on the contents (and sometimes motivation) of this commit, without reflecting the historical structure of the sub-commits that where squashed.

> I am not squashing further since this would hide that the original implementation is due to @agramfort

@agramfort can confirm but I don't think this is a problem as long as he is co-credited in the changelog.

mblondel commented:
@jmetzen Could you add me to calibration.py? I wrote the original calibration_curve code.

agramfort commented:
> @agramfort can confirm but I don't think this is a problem as long as he is co-credited in the changelog.

fine with me. Or just squash my first 4 commits into 1 so we both have
commits on this. No big deal anyway

Commit:
ENH Add probability calibration based on isotonic regr. and Platt's sigmoid fit + calibration-curve

CalibratedClassifierCV allows to calibrate the predicted probabilities of base classifiers based on a cross-validation scheme and either Platt's sigmoid fit or isotonic regression. This can be used to compensate for an under-confident or over-confident classifier. It allows also to turn the decision scores of a non-probabilistic classifier into valid probabilities.

The function calibration_curve allows to evaluate how well calibrated the probabilties returned by a classifier are. Ideally, the curve should be close to diagonal.

Commit:
ENH Brier-score loss metric for classifiers

The Brier score allows to evaluate the quality of the predicted probabilities of a classifier. It is defined as  the mean squared difference between (1) the predicted probability assigned to the possible outcomes for an item and (2) the actual outcome.

Commit:
TST Tests for the calibration module

Commit:
DOC Examples for the calibration of predicted probabilities and calibration-curves

Commit:
DOC Narrative doc for the calibration module

jmetzen commented:
Alright, I squashed it now into 5 generic commits (calibration module, brier score, tests, examples, narrative doc). I've also added @mblondel to the list of authors in calibration.py

agramfort committed:
COSMIT : pep8 + 2 spaces

agramfort committed:
TST improve coverage of calibration.py

agramfort commented:
I pushed a couple of commits (cosmit + coverage improvement). The coverage could be slightly improved. Currently we have 95% coverage.

ogrisel commented:
> Alright, I squashed it now into 5 generic commits (calibration module, brier score, tests, examples, narrative doc). I've also added @mblondel to the list of authors in calibration.py

Great, thank you very much!

ogrisel commented:
> The coverage could be slightly improved. Currently we have 95% coverage.

Indeed there is a couple of exceptions that should be covered by additional `assert_raises` or `assert_raise_message` checks:

https://coveralls.io/builds/1947924/source?filename=sklearn%2Fcalibration.py

It should be easy to raise the coverage close to 99% on that file.

jmetzen commented:
I've added some assert_raises, coverage of calibration.py should now be effectively 100%

ogrisel commented:
Thanks. I think this is ok for merge. @agramfort @mblondel any more comments?

agramfort commented:
+1 for merge on my side.

mblondel commented:
Just to confirm: do we want really want `sigmoid_calibration` to be a public function?

Other than that +1 as well!

agramfort commented:
Let's make sigmoid_calibration private indeed.

ogrisel commented:
+1 for a private `sigmoid_calibration` as well. I will do the change and merge.

arjoly commented on sklearn/metrics/classification.py:
```diff
@@ -1515,3 +1516,95 @@ def hinge_loss(y_true, pred_decision, labels=None, sample_weight=None):
+    return label_binarize(y_true, labels)[:, 0]
+
+
+def brier_score_loss(y_true, y_prob, sample_weight=None, pos_label=None):
```
Can you test this functions using the common tests in https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/metrics/tests/test_common.py ?

jmetzen replied:
Done (these tests are pretty nice; I didn't know they exist). I had to modify test_invariance_string_vs_numbers_labels() slightly such that pos_label is also set for THRESHOLDED_METRICS

arjoly commented on sklearn/metrics/classification.py:
```diff
@@ -1465,3 +1466,78 @@ def hinge_loss(y_true, pred_decision, pos_label=None, neg_label=None):
+    takes on a value between zero and one, since this is the largest
+    possible difference between a predicted probability (which must be
+    between zero and one) and the actual outcome (which can take on values
+    of only 0 and 1).
```
Tests are written for this use case in the common metric tests.

Commit:
TST Adding brier_score_loss to test_common.py

ogrisel commented:
Oops I had already merged, I will cherry pick this.

ogrisel commented:
Done! :beers: Thank you very much everyone!

agramfort commented:
yeah :beers: !!!

GaelVaroquaux commented:
:beers: indeed people. Congratulations!

jmetzen commented:
Thanks for merging! :beers:

mblondel commented:
Congrats!

When was this effort started again? Two? Three years ago? :)

amueller commented on examples/calibration/plot_calibration_curve.py:
This raises a warning as GaussianNB doesn't support sample_weights. If this is expected behavior, I think the warning should be caught.

agramfort commented on examples/calibration/plot_calibration_curve.py:
do we really want to catch warnings in examples... maybe just quickly add
sample_weights to GaussianNB? :)

amueller commented on examples/calibration/plot_calibration_curve.py:
Do we really want examples to raise warnings? ;) I thought the use of a classifier that doesn't have ``sample_weights`` was intentional.

agramfort commented on examples/calibration/plot_calibration_curve.py:
is it me or there is no use of sample_weight in this example?

amueller commented on examples/calibration/plot_calibration_curve.py:
I thought it was used internally in the calibration. But you are right, it shouldn't warn if it is not used. Will investigate!

amueller commented on examples/calibration/plot_calibration.py:
```diff
@@ -0,0 +1,116 @@
+
+# Gaussian Naive-Bayes with isotonic calibration
+clf_isotonic = CalibratedClassifierCV(clf, cv=2, method='isotonic')
+clf_isotonic.fit(X_train, y_train, sw_train)
```
my bad, this line is what caused the warning.

jmetzen replied:
i think we agreed that throwing a warning makes sense here: we can use the sample-weights during the calibration but not during the fitting of the base-classifier. We could catch the warning here but I agree with @agramfort that adding sample_weights to GaussianNB would be nicer.

agramfort replied:
can you point me to the warning raised? otherwise I need to dig into it

jmetzen replied:
line 151 in calibration.py in master [not included in this PR]

agramfort replied:
let's add sample weight to GNB then

jmetzen replied:
PR #4346 created that adds sample weights to GaussianNB
